### ISSUE TITLE
Export Private portion of the Key for CSR

### ISSUE THREAD

**Issue by jxu103** [tpm2-software/tpm2-tools#2133]
Hello, 

Is there a way to export the private portion of the AK from obj.priv (generated from creating the key) as RSA 2048 and pass it in to openssl for CSR? Currently I could not figure out a way for export the private portion of a key.

Thank you,
Jim

**Comment by idesai** [comment]
No, you cannot export a sensitive portion of the TPM generated key unless it is a duplicable key. What you can do is generate a key pair using OpenSSL and then import the private portion of the key in the TPM using [tpm2_loadexternal](https://github.com/tpm2-software/tpm2-tools/blob/master/man/tpm2_loadexternal.1.md)

**Comment by jxu103** [comment]
> No, you cannot export a sensitive portion of the TPM generated key unless it is a duplicable key. What you can do is generate a key pair using OpenSSL and then import the private portion of the key in the TPM using [tpm2_loadexternal](https://github.com/tpm2-software/tpm2-tools/blob/master/man/tpm2_loadexternal.1.md)

Thanks for the response! I have tried using it but i kept getting an error after the following command: 
tpm2_loadexternal -H n -r private.pem -c key.ctx
"ERROR: File "private.pem" size is larger than buffer, got 1679 expected less than 778"

**Comment by williamcroberts** [comment]
You could also use the openssl engine or pkcs11 engines to perform this activity. You'd probably have an easier time with the engine and this PR https://github.com/tpm2-software/tpm2-tss-engine/pull/140 since it will make importing existing keys into the engine easier. Im still working on that feature in tpm2-pkcs11.

**Comment by jxu103** [comment]
> You could also use the openssl engine or pkcs11 engines to perform this activity. You'd probably have an easier time with the engine and this PR [tpm2-software/tpm2-tss-engine#140](https://github.com/tpm2-software/tpm2-tss-engine/pull/140) since it will make importing existing keys into the engine easier. Im still working on that feature in tpm2-pkcs11.

Hello William,
Is it possible to use the tpm2_loadexternal" function? I am running into an error where the pem size is larger than buffer.

Thanks,
Jim

**Comment by idesai** [comment]
@jxu103, the following are the issues with your setup:

1. tpm2-tools v3.0.4 is really old and end of life, if you intend to do anything sustainable you should seriously consider moving to a 4.X branch.

2. tpm2-tools v3.0.4 does not support loading keys in PEM/ DER format. You will have to manually convert a PEM/ DER file to a TPM2B_PUBLIC structure and then use tpm2_loadexternal.

3. tpm2-tools v3.0.4 release did not have the tpm2_import tool which would have been another great option especially for a CSR like use case.

In short #2 is your best bet if you can't move to a higher version of the tools, but if you can move to 4.x, you absolutely should as these cases have been nicely handled there.

**Comment by jxu103** [comment]
> @jxu103, the following are the issues with your setup:
> 
> 1. tpm2-tools v3.0.4 is really old and end of life, if you intend to do anything sustainable you should seriously consider moving to a 4.X branch.
> 2. tpm2-tools v3.0.4 does not support loading keys in PEM/ DER format. You will have to manually convert a PEM/ DER file to a TPM2B_PUBLIC structure and then use tpm2_loadexternal.
> 3. tpm2-tools v3.0.4 release did not have the tpm2_import tool which would have been another great option especially for a CSR like use case.
> 
> In short #-2 is your best bet if you can't move to a higher version of the tools, but if you can move to 4.x, you absolutely should as these cases have been nicely handled there.

I agree that we should move to 4.x. Unfortunately we are using CentOS 7, I could be wrong but it seems like 4.x is not supported on CentOS 7, and need CentOS 8. 

I am trying to load a private key into the TPM, haven't been able to find anything for this.

**Comment by billatarm** [comment]
> @jxu103 @idesai @williamcroberts Hi, all, currently, we can use [tpm2_loadexternal](https://github.com/tpm2-software/tpm2-tools/blob/master/man/tpm2_loadexternal.1.md) load external RSA private key into tpm. However, there remains a strong demand for the CA to issue a certificate to AK, which requires using AK's private key to construct a CSR. Since AK's private key cannot be exported, how can we achieve this goal? One approach is to manually construct a CSR template and then sign it with AK's private key. This method doesn't seem to have much public documentation. Do you have any suggestions? Thank you.

You could persist the AK at an address and maybe use the openssl provider? @AndreasFuchsTPM or @gotthardp any other ideas?

**Comment by bronzeMe** [comment]
Thanks everyone, the [tpm2-tss-engine ](pm2-tss-engine](https://github.com/tpm2-software/tpm2-tss-engine)provides support for creating a CSR bound to the AK using OpenSSL-style commands.
